### PR TITLE
fix: prebid headerbidding on react pages

### PR DESCRIPTION
**Pull request by parkjh**
Previously headerbidding was not working on any react pages see https://www.thetimes.co.uk/edition/news/mays-brexit-backstop-is-barrier-to-new-trade-deals-says-leaked-legal-notes-tdbxtjx5h?react=1
this pr fixes the header bidding process by effectively passing the correct slot sizes

### CODE REVIEW

**Comment by parkjh** [comment]
@tuncaulubilge Hey Tunca, I was wondering if I could get some help today to debug react native on android?

File: packages/ssr/__tests__/cypress/support/index.js [hunk 1]
```diff
@@ -13,8 +13,4 @@
 // https://on.cypress.io/configuration
 // ***********************************************************
 
-// Import commands.js using ES2015 syntax:
 import "./commands";
-
```
Comment (colincclark) [inline packages/ssr/__tests__/cypress/support/index.js]: why were these removed?

File: packages/ssr/package.json [hunk 2]
```diff
@@ -15,10 +15,8 @@
     "start": "webpack --config=src/standalone-renderer/webpack.config.js && node src/standalone-renderer/app.js",
     "start:testservers": "node ./mock-server & wait-on tcp:4000 && GRAPHQL_ENDPOINT=http://localhost:4000/graphql yarn start & wait-on tcp:3000",
     "stop:testservers": "kill $(lsof -t -i:3000,4000)",
-    "cypress:debug": "DEBUG=cypress:* cypress open",
     "test:integration": "yarn bundle:prod && yarn start:testservers && cypress run; exitCode=$?; yarn stop:testservers; exit $exitCode",
-    "test:integration:dev": "yarn bundle:dev && yarn start:testservers && cypress run --headed; exitCode=$?; yarn stop:testservers; exit $exitCode",
```
Comment (colincclark) [inline packages/ssr/package.json]: why were these removed?

File: README.md [hunk 3]
```diff
@@ -104,7 +104,7 @@ Any of these source files can be debugged directly.
 To debug our native Storybook:
 
 1. `yarn storybook-native` and leave it running
-2. `yarn android` or `yarn ios`
+2. `yarn android:logs` or `yarn ios:logs`
```
Comment (colincclark) [inline README.md]: can you explain the difference in the README between these two commands rather than removing the `yarn android` and `yarn ios` script explanations please

File: android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt [hunk 4]
```diff
@@ -20,7 +20,7 @@ class ReactConfig : BaseJavaModule() {
             put("cookieEid", "")
             put("isLoggedIn", "")
             put("graphqlEndPoint", "https://api.thetimes.co.uk/graphql")
-            put("adNetworkId", "")
+            put("adNetworkId", "25436805")
```
Comment (colincclark) [inline android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: can this be added as a constant somewhere suitable?

File: package.json [hunk 5]
```diff
@@ -55,7 +57,6 @@
     "clean-snaps": "rm -rf dextrose/snappy/*.png",
     "commit": "git-cz",
     "test:e2e": "lerna run bundle && lerna run test:integration --stream",
-    "test:e2e:dev": "lerna run bundle && lerna run test:integration:dev --stream",
```
Comment (colincclark) [inline package.json]: why is this removed?

File: packages/ad/src/dom-context.web.js [hunk 6]
```diff
@@ -42,14 +43,19 @@ class DOMContext extends Component {
...
+    switch (type) {
+      case "renderFailed":
+        onRenderError();
+        break;
+      case "renderComplete":
+        onRenderComplete();
+        break;
+      default:
+        if (data.debug) {
+          logger(type, detail);
```
Comment (colincclark) [inline packages/ad/src/dom-context.web.js]: not sure about introducing debugging and error handling directly into the console?

File: packages/ad/src/utils/generate-config.js [hunk 7]
```diff
@@ -42,6 +42,7 @@ const getAdSizes = (sizeMap, width) => {
   return [];
 };
 
+// Returns the valid ad sizes given the ad slotName and window width
```
Comment (colincclark) [inline packages/ad/src/utils/generate-config.js]: this function does a little more than what the comment says, maybe lose the comment altogether?

File: packages/ad/src/utils/prebid-config.js [hunk 8]
```diff
@@ -35,8 +35,8 @@ const bidderSettings = ({ maxBid, minPrice, bucketSize }) => ({
   ]
 });
 
-const getPrebidSlotConfig = (pos, section, width, biddersConfig) => {
-  const sizes = getAdSizes(pos, width);
+const getPrebidSlotConfig = (slot, section, width, biddersConfig) => {
```
Comment (colincclark) [inline packages/ad/src/utils/prebid-config.js]: 👍

File: package.json (hunk 5 above)
Comment (parkjh) [reply package.json]: not useful anymore as there's debug mode for cypress

File: README.md (hunk 3 above)
Comment (parkjh) [reply README.md]: this outputs additional react-natives logs after it's built storybook native app

File: packages/ssr/package.json (hunk 2 above)
Comment (parkjh) [reply packages/ssr/package.json]: not useful anymore as there's debug mode for cypress

File: packages/ad/src/dom-context.web.js (hunk 6 above)
Comment (parkjh) [reply packages/ad/src/dom-context.web.js]: although it was previously logging, debug mode allows both the developer and commercial team to debug header bidding, which is very useful and difficult to do otherwise

File: README.md (hunk 3 above)
Comment (parkjh) [reply README.md]: commands without logs just builds the storybook native app without logging

File: android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt (hunk 4 above)
Comment (parkjh) [reply android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: not sure how to share config between react and kotlin...

Comment (colincclark) [reply android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: if it starts to block the PR - please create a github issue (tech debt)

**Comment by parkjh** [comment]
some of the tests were removed as the functionality itself was removed, as you may notice the removed tests were around throwing exception and error handling

**Comment by parkjh** [comment]
increased coverage to 91.07% https://coveralls.io/builds/20420144

File: README.md [hunk 9]
```diff
@@ -104,7 +104,7 @@ Any of these source files can be debugged directly.
 To debug our native Storybook:
 
 1. `yarn storybook-native` and leave it running
-2. `yarn android` or `yarn ios`
+2. `yarn android:logs` or `yarn ios:logs` (thie will build storybook app and output logs)
```
Comment (nehasri89) [inline README.md]: typo

File: android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt (hunk 4 above)
Comment (parkjh) [reply android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: will raise issue later

File: README.md [hunk 10]
```diff
@@ -104,7 +104,7 @@ Any of these source files can be debugged directly.
 To debug our native Storybook:
 
 1. `yarn storybook-native` and leave it running
-2. `yarn android` or `yarn ios`
+2. `yarn android:logs` or `yarn ios:logs` (this will build storybook app and output logs)
```
Comment (colincclark) [inline README.md]: still think the `yarn android` and `yarn ios` scripts should be documented

File: packages/ad/src/utils/logger.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+/* eslint-disable no-console */
+export default (type, detail) => {
```
Comment (aronshamash) [inline packages/ad/src/utils/logger.js]: can could be done without a switch statement `console[type](detail)`

File: android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt (hunk 4 above)
Comment (colincclark) [reply android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: can you raise an issue now so it does not get lost please?

File: packages/ad/__tests__/dom-context.native.js [hunk 12]
```diff
@@ -80,16 +51,76 @@ export default () => {
...
+    expect(console.warn).toHaveBeenCalledWith("message");
+
+    component.handleMessageEvent(makeMessageEvent("info", "message"));
+    expect(console.info).toHaveBeenCalledWith("message");
+
+    component.handleMessageEvent(makeMessageEvent("debug", "message"));
+    expect(console.debug).toHaveBeenCalledWith("message");
+  });
+
+  it("does not logs a console message when it receives a log event from the webview", () => {
```
Comment (colincclark) [inline packages/ad/__tests__/dom-context.native.js]: "does not log"

File: packages/ad/__tests__/utils/ad-init.js [hunk 13]
```diff
@@ -80,9 +80,10 @@ export default () => {
   it("throws if the init hook is called twice", () => {
     const init = adInit(initOptions);
     init.init();
-    expect(() => init.init()).toThrowError(
-      new Error("init() has already been called")
-    );
+    return init.init().catch(err => {
+      const expectedError = new Error("init() has already been called");
+      expect(err).toEqual(expectedError);
```
Comment (colincclark) [inline packages/ad/__tests__/utils/ad-init.js]: you can use https://jestjs.io/docs/en/expect#tothrowerror

File: packages/ad/src/utils/ad-init.js [hunk 14]
```diff
@@ -293,75 +371,80 @@ const adInit = args => {
...
-          if (amazonSlots.length > 0) {
+          try {
+            const amazonSlots = this.getAmazonConfig(
+              adsSlots,
+              networkId,
+              adUnit,
+              section
+            );
+            if (amazonSlots.length < 1) {
+              throw new Error("no ad slots are defined");
```
Comment (colincclark) [inline packages/ad/src/utils/ad-init.js]: "no amazon ad slots are defined" ?

File: packages/ad/src/utils/ad-init.js [hunk 15]
```diff
@@ -88,42 +37,94 @@ const adInit = args => {
...
+                slotName
+              );
+
+              if (!slot) {
+                throw new Error(
+                  `Ad slot ${slotName} ${adUnitPath} could not be defined, probably it was already defined`
+                );
+              }
+
+              slot.addService(window.googletag.pubads());
```
[the remark this answers is not in the document]
Comment (parkjh) [reply packages/ad/src/utils/ad-init.js]: @aronshamash I guess previously I took to the approach of if it ain't broke don't touch it, but agreed

**Comment by colincclark** [comment]
`ad-init` has quite a bit of missing coverage for the code you introduced. See coveralls link on the PR

File: packages/ad/__tests__/utils/ad-init.js (hunk 13 above)
Comment (parkjh) [reply packages/ad/__tests__/utils/ad-init.js]: @colincclark it doesn't throw Error though it returns promise reject, even though they are treated the same I think it make more sense this way...

File: packages/ad/src/utils/logger.js (hunk 11 above)
Comment (parkjh) [reply packages/ad/src/utils/logger.js]: @aronshamash agreed but I thought it would be better to be more verbose and it's easier for unit testing

**Comment by parkjh** [comment]
@aronshamash replied to your comments, let me know if you disagree...

File: packages/ad/src/utils/ad-init.js [hunk 16]
```diff
@@ -1,71 +1,21 @@
...
 // defined outside the function, so that it can be passed into a WebView.
-
 const adInit = args => {
   const { el, data, platform, eventCallback, window } = args;
   const { document, setTimeout, Promise } = window;
-
+  const enablePrebidding = platform === "web";
+  const hasBidInitialiser =
+    window.nuk && window.nuk.ads && window.nuk.ads.loaded;
+  const withoutHeaderBidding = enablePrebidding && !hasBidInitialiser;
```
Comment (nehasri89) [inline packages/ad/src/utils/ad-init.js]: `isBidInitialised` ?

Comment (parkjh) [reply packages/ad/src/utils/ad-init.js]: covered by `hasBidInitialiser` although I can rename that to `isBidInitialised` instead

**Comment by parkjh** [comment]
@jimhunty still failing after redeploying...

File: README.md (hunk 10 above)
Comment (colincclark) [reply README.md]: 2a. is not valid mark up - will just be weirdly indented

File: android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt (hunk 4 above)
Comment (colincclark) [reply android/app/src/main/java/com/storybooknative/stubs/ReactConfig.kt]: can you add the issue link to the PR?
